File: 225-Implement Stack using Queues-用队列实现栈/225-Implement Stack using Queues-用队列实现栈.java
class MyStack {

	private Queue<Integer> queue1;
	private Queue<Integer> queue2;
	
    public MyStack() {
    	queue1 = new PriorityQueue<>();
    	queue2 = new PriorityQueue<>();
    }
    
    public void push(int x) {
    	queue1.add(x);
    }
    
    private int reQueue(boolean pop) {
    	Integer preVal = null;
    	while(!queue1.isEmpty()) {
    		if(preVal != null) queue2.add(preVal); 
    		preVal = queue1.poll();
    	}
    	if(!pop) {
    		queue2.add(preVal);
    	}
    	Queue<Integer> temp = queue1;
    	queue1 = queue2;
    	queue2 = temp;
    	return preVal;
    }
    
    public int pop() {
    	return reQueue(true);
    }
    
    public int top() {
    	return reQueue(false);
    }
    
    public boolean empty() {
    	return queue1.isEmpty();
    }
}

/**
 * Your MyStack object will be instantiated and called as such:
 * MyStack obj = new MyStack();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.top();
 * boolean param_4 = obj.empty();
 */

File: 225-Implement Stack using Queues-用队列实现栈/225-Implement Stack using Queues-用队列实现栈.js
var MyStack = function() {
    this.queue1 = []
    this.queue2 = []
};



/** 
 * @param {number} x
 * @return {void}
 */
MyStack.prototype._revert = function(pop) {
    let preVal = null
    while(this.queue1.length !== 0){
        if(preVal !== null){
            this.queue2.unshift(preVal)
        }
        preVal = this.queue1.pop()
    }
    if(!pop){
        this.queue2.unshift(preVal)
    }
    let temp = this.queue1
    this.queue1 = this.queue2
    this.queue2 = temp
    return preVal
};

/** 
 * @param {number} x
 * @return {void}
 */
MyStack.prototype.push = function(x) {
    this.queue1.unshift(x)
};

/**
 * @return {number}
 */
MyStack.prototype.pop = function() {
    return this._revert(true)
};

/**
 * @return {number}
 */
MyStack.prototype.top = function() {
    return this._revert(false)
};

/**
 * @return {boolean}
 */
MyStack.prototype.empty = function() {
    return this.queue1.length === 0
};

/**
 * Your MyStack object will be instantiated and called as such:
 * var obj = new MyStack()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.empty()
 */